README.md


# mc-catalog

This repository contains the set of scripts built to analyze and plot the
results of the **ASteCA** processing of the 239 Magellanic Cloud clusters in our
database, observed with `C, T1` Washington photometry.

The main script is `mc_cat_analysis.py`. It produces all the figures and data
used in the article.

Below we present a description of every file/folder found in this repository.

- [Top level](#top-level)
  - [`AMRs/`](#AMRs)
  - [`aux_funcs/`](#aux_funcs)
  - [`databases/`](#databases)
  - [`extinction_MCEV/`](#extinction_mcev)
  - [`figures/`](#figures)
  - [`functions/`](#functions)
  - [`runs/`](#runs)

## Top level

* `README.md`

 This file.

* `asteca_output_final.dat`

 Combined final fundamental parameters for the entire sample, from the analysis
 performed with **ASteCA**.

* `lit_OCs_data.ods`

 Literature data on each cluster.

* `mc_cat_analysis.py`

 Script that Produces the main figures used in the article. Functions are stored
 in the `functions/` folder.

* `requirements.txt`

 Requirements to run the scripts in this repository. Install with:
 `pip install -r requirements.txt`.


### `AMRs/`

  Data files with literature AMR values.


### `aux_funcs/`

* `add_p_vals/`

  Contains the script and input/output files used to add KDE p_values
  to the `asteca_output.dat` file from the 1st run, since the function was
  turned off.

* `comp_metals.py`

 Compares real versus rounded metallicity values, to asses the impact of an
 issue with the rounding function of **ASteCA**.

* `move_files_names.py`

 Script to move .png files from their `input_XX/` folders for each run, into
 a single folder containing all the analyzed clusters.
 This script uses the `asteca_output_XX.dat` files.

* `move_files_sizes.py`

 Script to move files into `input_XX/` folders distributed so that each
 folder has approximately the same total size.

* `photom_error_TO.py`

 Reads all cluster data files, and plots their T1 and (C-T1) photometric errors
 around the turn-off region.

* `print_runs_clust_values.py`

 Prints the final parameters found in each run for a given cluster.


### `databases/`

* `age-mass_relations.ods`

 Ages and masses from the articles [Piskunov et al. (2008)]
 (http://adsabs.harvard.edu/abs/2008A%26A...477..165P) and [Baumgardt
 et al. (2012)](http://arxiv.org/abs/1207.5576)

* `BB_10.dat`

 List of ages for 151 SMC and 539 LMC clusters in the database presented in
 [Bonato & Bica (2010)](http://cdsads.u-strasbg.fr/abs/2010MNRAS.403..996B).
 Only 285 out of the 690 clusters have an age value assigned.

* `bb_cat.dat`

 RA & DEC positions for the 3740 clusters in the
 [Bica et al. (2008)](http://cdsads.u-strasbg.fr/abs/2008MNRAS.389..678B)
 catalog ([Table 3](http://vizier.u-strasbg.fr/viz-bin/VizieR-3?-source=J/
 MNRAS/389/678/table3)).

* `chiosi_06.dat`

 Contains 311 SMC clusters younger than 1 Gyr (ages determined using isochrone
 fitting) from the work by [Chiosi et al. (2006)](http://adsabs.harvard.edu/
 abs/2006A%26A...452..179C). Also lists several other structures/objects, ie:
 association, supernovas, H II regions and entities in between.

* `cross_h03_p12.py`

 Matches the data in the H03 and P12 databases.

* `cross_match.py`

 Matches the data in the above databases to those clusters processed by
 ASteCA.

* `glatt_10.dat`

 CMD ages taken from the [Glatt et al. (2010)](http://www.aanda.org/10.1051/
 0004-6361/201014187) catalog for 1194 LMC clusters and 322 SMC clusters.

* `H03_P12_masses.ods`

 Masses given in the H03 and P12 articles.

* `hunter_03.dat`

  Clusters in the S/LMC from [Hunter et al. (2003)](http://adsabs.harvard.edu/
  abs/2003AJ....126.1836H), 748 belong to the LMC and 191 to the SMC for a
  total of 939 clusters.

* `Maia_13_mass.dat`

 [Maia et al. (2013)](http://adsabs.harvard.edu/abs/2013arXiv1310.5934M) mass
 distribution and structural parameters of 29 Small Magellanic Cloud star
 clusters.

* `matched_clusters.dat`

 Output of the `cross_match.py` script.  Final data file with all matched
 clusters between **ASteCA** and the databases in the literature.

* `matched_clusters_G10.dat`

 Same as `matched_clusters.dat`, but using the G10 database instead of the
 **ASteCA** database as the reference to match.

* `matched_H03_P12.dat`

 Output of the `cross_h03_p12.py` script. Crossed-matched OCs from H03 and P12.

* `McL_vdM05.dat`

 Masses from [McLaughlin & van der Marel (2005)](http://adsabs.harvard.edu/
 abs/2005ApJS..161..304M). Only three OCs could be cross matched: LMC-NGC1860,
 LMC-SL663, and SMC-NGC339.

* `pietrz_00_LMC.dat`

  Ages for 600 LMC clusters obtained in [Pietrzynski & Udalski (2000)](http://
  adsabs.harvard.edu/abs/2000AcA....50..337P).

* `pietrz_99_SMC.dat`

 Ages for 93 SMC clusters obtained in [Pietrzynski & Udalski (1999)](http://
 adsabs.harvard.edu/abs/1999AcA....49..157P).

* `popescu_12_LMC.dat`

 List of ages and masses for 632 clusters in the database presented in
 [Popescu et al. (2012)](http://adsabs.harvard.edu/abs/2012ApJ...751..122P)
 with ages correlated to the *Hunter et al* catalog.

* `popescu_12_LMC_glatt.dat`

 Idem above, but also adds ages taken from the *Glatt et al.* catalog, for
 288 clusters in the *Hunter et al.* catalog.

* `rafelski_05_SMC.dat`

  Ages for 195 SMC clusters from [Rafelski & Zaritsky (2005)](http://
  adsabs.harvard.edu/abs/2005AJ....129.2701R), obtained via
  integrated photometry.


### `extinction_MCEV/`

* `cls_exts_match.dat`

 Output of the `extin_analysis.py` script.

* `extin_analysis.py`

 Gets the extinction data from the table obtained via the MCEV service
 (`ra_dec_exts_mult_matches.dat`) and produces for each of the clusters a
 value of the closest, average and maximum E(B-V) extinction values.
 The transformation equation used is:

 `E(V-I) = 1.38 * E(B-V)`

 according to [Tammann et al. (2003)]
 (http://adsabs.harvard.edu/abs/2003A%26A...404..423T). The results are stored
 in  the `cls_exts_match.dat` file.

* `IRSA_MC_ext.tbl`

  Output of the [IRSA](http://irsa.ipac.caltech.edu/applications/DUST/) query to
  obtain the [Schlafly & Finkbeiner
  (2011)](http://adsabs.harvard.edu/abs/2011ApJ...737..103S) corrected `E(B-V)`
  values for all the clusters.

* `IRSA_BB_ext.tbl`

  Idem above for the 3740 clusters in the Bica et al. catalog.

* `ra_dec.dat`

 RA & DEC data for all clusters in CSV format.

* `ra_dec_exts_mult_matches.dat`

 Output of the MCEV service query with TOPCAT. Each of the cluster's position is
 matched with at least one area in the reddening maps containing a E(V-I) value.

 The cluster [OHSC28](http://simbad.u-strasbg.fr/simbad/sim-id?Ident=OHSC+28)
 needed a 6.0 deg search radius to find areas with extinction in the maps,
 making it the one furthest away from the areas studied in the MCEV maps.

* `TOPCAT_instruct.dat`

 Instructions to perform a
 [MCEV](http://dc.zah.uni-heidelberg.de/mcextinct/q/cone/form) query using
 TOPCAT.

* `topcat-lite.jnlp`

 File that downloads and runs the lite version of TOPCAT (see instructions to
 run).


### `figures/`

 Output figures from main script.


### `functions/`

 `*.py`: functions called by the main script.

 `0.004.dat` & `0.008.dat`: Marigo isochrones used by the
 `CMD_obs_vs_asteca.py` script to plot the G10 fitted isochrones.

 `mar2008_ubvrijhk/` & `parsec11_ubvrijhk/`: folders that contain the Marigo
 and PARSEC isochrones compared by the `marigo_parsec_isochs.py` script.


### `OCs_data/`

 This folder contains data necessary for the scripts to produce the analysis and
 the plots.


### `runs/`

 Folder that contains the results for all the runs.

* `large_diff_age_clusts.ods`: analysis of the obtained values in different runs
 for those clusters that display a large age difference with the literature
 values.

#### `1st_run/`

  First batch of data obtained using the following parameters:

  - Semi mode.
  - Center found with 100px search area.
  - Auto radius.
  - Field regions: 10.
  - Decontamination algorithm: auto.
  - Reduced membership: Bayesian blocks binning
  - Restricted range in extinction: MCEV max + 0.1.
  - Restricted distance modulus:  SMC, [18.8, 19.201, 0.05] ;
    LMC, [18.3, 18.701, 0.05].
  - Best fit: Dolphin + Knuth binning.

#### `2nd_run/`

  Second batch of data for 83 clusters (60 LMC, 23 SMC) from the 1st run,
  marked to be re-processed. The data was obtained using the same extinction
  and distance modulus range, and changing the following parameters:

  * Center, radius, number of field regions, max E(B-V) and binning of red_memb
   and best_fit, as described in `README.dat`.

  The following clusters were left un-processed (no reason, just didn't bother):

  * NGC1917, SL579, SL588, LW54, NGC419, SL244

#### `3rd_run/`

  Third batch of data obtained using the following parameters, for all
  clusters:

  * Semi mode with center and radius fixed for several clusters.
  * Number of field regions set individually for some clusters.
  * Restricted range in extinction: **MCEV max for all clusters**.
  * Restricted distance modulus:  SMC, [18.86, 19.061, 0.02] ;
    LMC, [18.4, 18.601, 0.02].

#### `4th_run/`

  Fourth run, 44 clusters from 3rd run re-processed. Same parameters as in
  3rd run but increased number of generations and mutation rate, and skipped
  the removal of stars post-DA.
  Some clusters had their parameters adjusted in `semi_input.dat`.

  * skip red_memb
  * gens=3000
  * mut_rate=0.2

#### `5th_run/`

  25 clusters from 4th run re-process. Increased mutation rate, three blocks
  of clusters processed with the following options:

  * mut_rate=0.25 (all blocks)
  * skip DA + red_memb: local + scott
  * skip DA + red_memb: local + blocks
  * skip DA + skip red_memb <-- Will fit mostly field stars if n_memb is low.

  For NGC339 (input_05) the max mass limit was set to 30000.

#### `6th_run/`

  Sixth run, 16 problematic clusters processed in three blocks with:

  * DA + red_memb=scott
  * skip DA + red_memb=scott
  * DA + red_memb=blocks (same as 3rd run)

#### `7th_run/`

  Seventh run, 20 high mass clusters re-processed with higher mass max limit.
  Parameters used:

  * max mass = 30000
  * n_gen = 3000
  * mut = 0.25

#### `8th_run/`

  Re-process 8 clusters from 1st and 2nd run with extinction and distance
  modulus limits as in 3rd run, more generations and an increased mutation
  rate.

  * Restricted range in extinction: MCEV max for all clusters.
  * Restricted distance modulus:  SMC, [18.86, 19.061, 0.02] ;
    LMC, [18.4, 18.601, 0.02]
  * n_gen = 3000
  * mut = 0.25

#### `9th_run/`

  Three clusters re-processed from the 8th run that showed large age
  differences, with the following parameters changed:

  * n_gen = 5000
  * n_el = 10
  * n_ei = 100
  * n_es = 25

#### `10th_run/`

  Re-process 13 cluster. Two are the ones were parameter values are still
  taken from the 2nd run, and the remaining 11 are those with
  `\delta log(age)>= 0.5`.

  Use the following parameters:

  * DA + red_memb = scott
  * n_gen = 5000
  * n_el = 1
  * n_ei = 25
  * n_es = 150

#### `11th_run/`

 Re-process the 2 clusters still dragged from the 2nd run (SL218 & BSDL654)
 plus 4 more clusters with large age differences (H88-131, BSDL631, L35, SL579).

 Use the following parameters:

 * DA + red_memb = different values per cluster
 * n_gen = 3000
 * p_mut = 0.1
 * n_el = 1
 * n_ei = 50
 * n_es = 30

#### `12th_run/`

 Re-process SL218 using the following parameters:

 * field regions = 25
 * DA + red_memb = auto + local (blocks)
 * n_pop = 500
 * n_gen = 2000
 * p_mut = 0.5
 * n_el = 1
 * n_ei = 25
 * n_es = 70

#### `13th_run`

 Re-process all clusters using a binary fraction of 0.

 Eight clusters from the SMC are missing (got lost somewhere in the FTP moving).
 These clusters will not be present in the runs: 14, 15, 16.

#### `14th_run`

  Re-process all clusters using fixed distance moduli (18.49/18.96, S/LMC) and
  increasing the maximum extinction by 0.05. Maximum mass increased to 30000.
  * No bootstrap process is applied.

#### `15th_run`

  Re-process all clusters using:

  * No bootstrap.
  * Range in extinction: **MCEV max + 0.1** for all clusters.
  * Restricted distance modulus:  SMC, [18.86, 19.061, 0.02] ;
    LMC, [18.4, 18.601, 0.02]
  * Maximum mass: 30000.

#### `16th_run`

  Equal to 15th run but using a fixed mass of 1500 Mo and `Tolstoy` likelihood
  instead of `Dolphin`.
  **Correction**: I used  *without noticing* a mass interval of:
  `TM 10 1500` in the `params_input.dat` file instead of fixing it to a value
  of `1500`.
  This caused the mass values in the range to be defined by
  `np.arange(10, 1500)` as `[  10,   11,   12, ..., 1497, 1498, 1499]`, ie:
  a step of 1 solar mass.

#### `17th_run`

 Process 8 clusters with large age differences: BSDL631, H88-131, H88-316,
 KMHK975, KMHK979, SL218, SL579, L35.

 * Range in extinction: **MCEV max** for all clusters.
 * Restricted distance modulus:  SMC, [18.86, 19.061, 0.02] ;
   LMC, [18.4, 18.601, 0.02]
 * Fixed mass: 1500.
 * `Tolstoy`.

#### `18th_run`

 *Important*: for all these clusters' data files the x,y columns are inverted.

 Process the 29 SMC clusters from Maia et al. (2014). Use the same parameters
 as in the 3rd run:

 * Semi mode with center and radius fixed for several clusters.
 * Number of field regions set individually for some clusters.
 * Restricted range in extinction: **MCEV max for all clusters**.
 * Restricted distance modulus:  SMC, [18.86, 19.061, 0.02] ;
   LMC, [18.4, 18.601, 0.02].
 * Total mass: 10000

Per cluster DA/GA parameters:

##### 1st run (29)
 1. 3000 generations
  * H86-97 (crashed, got not results)
 1. DA: auto + local + knuth
  * B103, H86-190, K47, NGC241, NGC242
 1. DA: auto + local + scott
  * B99
 1. DA: skip + local + knuth
  * B48, H86-76, H86-85, K61, SOGLE196
 1. DA: skip + local + scott
  * B55, H86-87, H86-90
 1. DA: skip + local + blocks
  * B134

###### Clusters kept from 1st run (12 + H86-188)
 * B103, B111, BS75, BS80, H86-76, H86-174, HW52, K43, K55, K57, K63, SOGLE196,
 H86-188 (it is processed several more times, but this one is kept)

##### 2nd run (17)
1. Defaults (DA: auto + local + blocks)
 * B124 (03), B55 (06), B99 (07), H86-188 (11), HW32 (18), NGC241 (27)
1. DA: skip + local + knuth
 * B134 (04)
1. DA: auto + local + knuth
 * B48 (05), H86-85 (14), H86-87 (15), H86-90 (16), H86-97 (17), K61 (24)
1. DA: auto + local + sqrt
 * H86-190 (12)
1. DA: auto + skip
 * K47 (21), L39 (26), NGC242 (28)

###### Clusters kept from 2nd run (6)
 * B55, B99, HW32, L39, K61, K47

##### 3rd run (11)
**All**: p_mut=0.35
1. Defaults params for DA
 * B48 (05), B124 (03), H86-188 (11), NGC241 (27), H86-97 (17), H86-190 (12),
   NGC242 (28)
1. DA: skip + local + knuth
 * H86-85 (14), H86-90 (16)
1. DA: skip + local + scott
 * B134 (04)
1. DA: skip + local + bb
 * H86-87 (15)

###### Clusters kept from 3rd run (2)
 * H86-190, NGC241

##### 4th run (9 + NGC241 is run again but 3rd run is kept)
**All**: p_mut=0.5, n_ei=25, n_es=60
1. Defaults params for DA
 * B124 (03), B134 (04), B48 (05), H86-188 (11), H86-87 (15)
1. DA: auto + local + knuth
 * H86-85 (14)
1. DA: auto + skip
 * H86-90 (16), H86-97 (17), NGC241 (27), NGC242 (28)

###### Clusters kept from 4th run (3)
 * B48, H86-87, B124

##### 5th run (6)
**All**: n_gen=3000, p_mut=0.5, n_ei=50, n_es=40
1. Defaults params for DA
 * H86-188 (11), H86-90 (16)
1. DA: auto + local + knuth
 * B134 (04)
1. DA: auto + skip
* H86-85 (14), H86-97 (17)
1. DA: skip + local + knuth
 * NGC242 (28)

###### Clusters kept from 5th run (4)
 * H86-85, NGC242, H86-97, B134

##### 6th run (3, B134 is processed again)
1. Defaults params for DA; GA: n_pop=200, n_gen=2000, p_mut=0.85, n_el=5
 * B134 (04), H86-188 (11), H86-90 (16)

###### Clusters kept from 6th run (1)
 * H86-90

##### 7th run (1)
1. Defaults params for DA; GA: n_pop=50, n_gen=2000, p_mut=0.5, n_el=1
 * H86-188 (11)

##### 8th run (1)
1. Defaults params for DA; GA: n_pop=200, n_gen=2000, p_mut=0.95, n_el=10
 * H86-188 (11)

##### 9th run (1)
1. DA: skip + local + knuth; GA: n_pop=100, n_gen=2000, p_mut=0.35, n_el=1
 * H86-188 (11)

##### 10th run (8)
Re-run 8 clusters that needed an offset applied to their photometry (which I
found out after processing them in the runs above): B48, H86-76, H86-90,
H86-97, L39, NGC241, NGC242, SOGLE196.
1. p_mut=0.35: NGC241
2. p_mut=0.5, DA skip, local+knuth: H86-76; p_mut=0.5, DA auto, skip: H86-97
3. p_muy=0.5, n_gen=3000, 

#### `19th_run`

Re-process these clusters to try to improve their fits. H88-131 could be better
fitted for the age and the other two for the metallicity.

1. DA: skip + local + blocks; GA: p_mut=0.35
* H88-131
1. DA: defaults; GA: p_mut=0.35
* NGC294, HW85

#### `20th_run`

Re-process L45 and L50 with better radii estimates. Originally, the values used
where:
rad_L45 = 162.5 (semi); rad_L50 = 170. (auto)

new values are:
rad_L45 = 100. (auto); rad_L50 = 100 (semi)

Since no substantial change is made, these new values are incorporated into the
3rd and 4th run where L50 and L45 where processed, respectively.
KDE probs are borrowed from the previous run, since they were not obtained
here.

Re-process NGC294 and HW85 fixing all parameters except metallicity, with
binary fractions of [0., 0.25, 0.5, 0.75, 1.]
Metallicity of NGC294 changed 0.001 --> 0.0005, due to issue #248.
These runs are only used to check the metal-binarity dependence, and are not
added to the final output file.

#### `21st_run`

Re-process 5 OCs whose masses differ greatly from the H03 and P12 estimates.
We use:

* no DA + no local removal of field stars
* no bootstrap
* max mass limit of 500000 with a step of 5000 (Mo)

All other parameter ranges are kept from 3rd run.

#### `22nd_run`

Re-process NGC 419 with a larger radius: 310px in 21st run vs 460px here. The
largest mass value found (with no field star cleaning applied) is 9e04.

#### `23rd_run`

Process 768 synthetic MASSCLEAN clusters to validate ASteCA's mass estimation.



article/CHANGELOG.md

# Article

*Astrophysical properties of star clusters in the Magellanic Clouds
homogeneously estimated by ASteCA*, Perren et al. (2017)


## Versions

#### v1.3 - 2017/04/21

* Corrections made on proof version.


#### v1.2 - 2017/02/14

* Version uploaded to SAGA (https://saga.edpsciences.org). All figures were
brought to the root folder.


#### v1.1 - 2017/02/09

* Version with comments removed.


#### v1.0 - 2017/02/07

* Version with language corrections accepted and commented.


#### v0.9_arxiv - 2017/01/30

* Version uploaded to arXiV. All images are brought to the top folder,
and reduced below 50% size.


#### v0.9 - 2017/01/27

* Changes from Rev 3 accepted by the editor. Removed some left over garbage and
commented lines, fixed a couple of minor things, added a 2 to the likelihood
equation.
* Sent for language corrections.


#### v0.8 - 2017/01/26

* Revision 3. Changes requested by the Editor, plus a correction in the
explanation of the effect of crowding in our analysis.


#### v0.7 - 2016/12/29

* Revision 2. Version with second batch of referee corrections implemented.


#### v0.6 - 2016/11/01

* Revision 1. Version with first batch of referee corrections implemented.


#### v0.5 - 2016/08/12

* Corrections by Vazquez applied. Sent to Piatti for next round of corrections.
* Version uploaded to A&A.


#### v0.4 - 2016/08/04

* Second half of the article corrected by Piatti.
* Sent to Piatti for final corrections for this 1st round.


#### v0.3 - 2016/08/01

* First half of the article corrected by Piatti. The corrections were
implemented into the article or rejected. Comments are shown in the
'changes_cmmts.txt' file.
* Sent to Piatti for 1st round of corrections, 2nd half of the article.


#### v0.2 - 2016/07/18

* Template changed to A&A.
* Many modifications (primarily added mass validation using MASSCLEAN synthetic
clusters)
* Sent to Piatti for 1st round of corrections.


#### v0.1 - 2016/06/21

* Initial version of the manuscript, written using MNRAS's LaTeX template.


functions/get_params.py

import numpy as np
from functions.photom_dispersion import get_disp


def float_str(val):
    '''
    Return float depending on whether the value is a float or a string.
    '''
    try:
        val_f = float(val)
    except:
        val_f = -9999999999.9

    return val_f


def rad_in_pc(float_lst):
    '''
    Convert radius from pixel to arcsec to parsecs.
    '''
    # Unpack
    r_px, scale, dis_mod, e_bv = float_lst
    # Convert from px to arcsecs.
    rad_arcsec = r_px * scale
    Av = 3.1 * e_bv
    # Distance to cluster in parsecs.
    d_pc = 10 ** (0.2 * (dis_mod + 5 - Av))
    # Radius in parsecs.
    r_pc = d_pc * np.tan(np.deg2rad(rad_arcsec / 3600.))

    return r_pc


def correct_int_col_extin(int_col, extinc):
    '''
    Correct integrated color for the extinction.
    '''
    E_CT1_E_BV = 1.97
    int_col_cor = int_col - E_CT1_E_BV * float_str(extinc)

    return int_col_cor


def z_to_feh(z, ez):
    '''
    Convert z to [Fe/H] for ASteCA values.
    '''
    # Use minimum metallicity value if z=0.
    z = max(0.0001, z)
    fe_h = np.log10(z / 0.0152)
    e_fe_h = (1. / np.log(10.)) * (ez / z)
    # Trim error if it's too large.
    e_fe_h = min(e_fe_h, 2.)

    return fe_h, e_fe_h


def params(r_path, as_names, as_pars, cl_dict, names_idx):
    '''
    Return ASteCA output and literature parameters values.
    '''
    # Indexes of columns in ASteCA output file.
    a_zi, a_zei, a_ai, a_aei, a_ei, a_eei, a_di, a_dei, a_mi, a_mei, a_rad, \
        a_erad, a_int_c, a_nmemb, a_CI, a_prob, a_r_core, a_e_r_core =\
        19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 4, 5, 18, 13, 11, 17, 6, 7

    # Indexes of columns in .ods literature file.
    ra_i, dec_i, gal_i, l_zi, l_zei, l_ai, l_aei, l_ei, l_eei, l_di, l_dei, \
        l_rad, l_scale, l_e_sandf, l_e_e_sandf, l_e_mcev, l_e_mcev_max, \
        l_e_e_mcev, l_mcev_dist, l_mass, l_e_mass = \
        cl_dict[0].index(u'ra_deg'), cl_dict[0].index(u'dec_deg'), \
        cl_dict[0].index(u'Galaxia'), cl_dict[0].index(u'[Fe/H] (dex)'), \
        cl_dict[0].index(u'e_Fe/H'), cl_dict[0].index(u'log(age)'), \
        cl_dict[0].index(u'e_log(age)'), cl_dict[0].index(u'E(B-V) (lit)'), \
        cl_dict[0].index(u'e_E(B-V)'), cl_dict[0].index(u'(m-M)o (mag)'), \
        cl_dict[0].index(u'e_(m-M)o'), cl_dict[0].index(u'rad (eye)'), \
        cl_dict[0].index(u'arcsec/pixel'), cl_dict[0].index(u'E_B_V_SandF'), \
        cl_dict[0].index(u'stdev_E_B_V_SandF'), \
        cl_dict[0].index(u'E_BV_closer_MCEV'), cl_dict[0].index(u'E_BV_max'), \
        cl_dict[0].index(u'E_BV_std_dev'), cl_dict[0].index(u'Dist (deg)'), \
        cl_dict[0].index(u'Mass'), cl_dict[0].index(u'e_mass')

    # Initialize empty lists. The first sub-list in the parameters list
    # corresponds to clusters in the SMC and the second to those in the LMC.
    gal_names, int_colors, n_memb, cont_ind, kde_prob, ra, dec, rad_pc, \
        erad_pc, r_core_pc, e_r_core, phot_disp = \
        [[[], []] for _ in range(12)]
    # First sub-list stores SMC values, the second one stores LMC values.
    # First and 2nd sub-sublist store Schlafly & Finkbeiner extinction values
    # and their errors. Third and 4th store MCEV extinction values and their
    # errors.
    ext_sf = [[[], []], [[], []]]
    ext_mcev = [[[], [], [], []], [[], [], [], []]]
    # First sub-list stores ASteCA values, the second one stores literature
    # values.
    zarr, zsigma, aarr, asigma, earr, esigma, darr, dsigma, marr, marr, \
        msigma, rarr = ([[[], []], [[], []]] for i in range(12))

    for i, as_p in enumerate(as_pars):

        # j = 0 for SMC clusters and 1 for LMC.
        j = 0 if cl_dict[names_idx[i]][gal_i] == 'SMC' else 1

        # Store names for each galaxy.
        gal_names[j].append(as_names[i])
        # Store coordinates.
        ra[j].append(cl_dict[names_idx[i]][ra_i])
        dec[j].append(cl_dict[names_idx[i]][dec_i])
        # Integrated color.
        int_col_no_corr = float_str(as_p[a_int_c])
        # Correct for extinction.
        int_col_corr = correct_int_col_extin(int_col_no_corr, as_p[a_ei])
        # Store extinction corrected integrated color.
        int_colors[j].append(int_col_corr)
        # Approx number of members.
        n_memb[j].append(float_str(as_p[a_nmemb]))
        # Contamination index.
        cont_ind[j].append(float_str(as_p[a_CI]))
        # Cluster KDE p-value probability.
        kde_prob[j].append(float_str(as_p[a_prob]))

        # Store literature E(B-V) values: Schlafly & Finkbeiner (SandF) and
        # MCEV.
        ext_sf[j][0].append(cl_dict[names_idx[i]][l_e_sandf])
        ext_sf[j][1].append(cl_dict[names_idx[i]][l_e_e_sandf])
        ext_mcev[j][0].append(cl_dict[names_idx[i]][l_e_mcev])
        ext_mcev[j][1].append(cl_dict[names_idx[i]][l_e_mcev_max])
        ext_mcev[j][2].append(cl_dict[names_idx[i]][l_e_e_mcev])
        ext_mcev[j][3].append(cl_dict[names_idx[i]][l_mcev_dist])

        # Store radius value in parsecs.
        float_lst = []
        for el in [as_p[a_rad], cl_dict[names_idx[i]][l_scale], as_p[a_di],
                   as_p[a_ei]]:
            # Store in list as floats.
            float_lst.append(float_str(el))
        r_pc = rad_in_pc(float_lst)
        rad_pc[j].append(r_pc)
        # Repeat process for errors in radius.
        float_lst = []
        for el in [as_p[a_erad], cl_dict[names_idx[i]][l_scale], as_p[a_di],
                   as_p[a_ei]]:
            # Store in list as floats.
            float_lst.append(float_str(el))
        e_r_pc = rad_in_pc(float_lst)
        erad_pc[j].append(e_r_pc)
        # Calculate r_core and its error in parsecs, using a simple 3 rule.
        r_core_px = float_str(as_p[a_r_core])
        e_r_core_px = float_str(as_p[a_e_r_core])
        px_2_pc_scale = r_pc / float_str(as_p[a_rad])
        if r_core_px > 0.:
            r_c_pc = r_core_px * px_2_pc_scale
            e_r_c_pc = e_r_core_px * px_2_pc_scale
        else:
            # Dummy values for clusters with no r_core values.
            r_c_pc, e_r_c_pc = -10., 0.
        # Store core radius and its error in parsecs.
        r_core_pc[j].append(r_c_pc)
        e_r_core[j].append(e_r_c_pc)
        # if r_c_pc > r_pc:
        #     print j, as_names[i], as_p[a_rad], r_core_px

        # Get photometric dispersion parameter.
        phot_disp[j].append(get_disp(r_path, as_names[i]))

        # Organize param values, ASteCA first, lit second.
        met = [as_p[a_zi], cl_dict[names_idx[i]][l_zi]]
        smet = [as_p[a_zei], cl_dict[names_idx[i]][l_zei]]
        age = [as_p[a_ai], cl_dict[names_idx[i]][l_ai]]
        sage = [as_p[a_aei], cl_dict[names_idx[i]][l_aei]]
        ext = [as_p[a_ei], cl_dict[names_idx[i]][l_ei]]
        sext = [as_p[a_eei], cl_dict[names_idx[i]][l_eei]]
        dis = [as_p[a_di], cl_dict[names_idx[i]][l_di]]
        sdis = [as_p[a_dei], cl_dict[names_idx[i]][l_dei]]
        mass = [as_p[a_mi], cl_dict[names_idx[i]][l_mass]]
        smass = [as_p[a_mei], cl_dict[names_idx[i]][l_e_mass]]
        rads = [as_p[a_rad], cl_dict[names_idx[i]][l_rad]]

        # Store ASteCA values (k=0) and literature values (k=1).
        for k in [0, 1]:

            if k == 0:
                z, ez = float_str(met[k]), float_str(smet[k])
                # Convert z to [Fe/H] for ASteCA values.
                fe_h, e_fe_h = z_to_feh(z, ez)
                zarr[j][k].append(fe_h)
                zsigma[j][k].append(e_fe_h)
            else:
                zarr[j][k].append(float_str(met[k]))
                zsigma[j][k].append(float_str(smet[k]))

            aarr[j][k].append(float_str(age[k]))
            asigma[j][k].append(float_str(sage[k]))
            earr[j][k].append(float_str(ext[k]))
            esigma[j][k].append(float_str(sext[k]))
            darr[j][k].append(float_str(dis[k]))
            dsigma[j][k].append(float_str(sdis[k]))
            marr[j][k].append(float_str(mass[k]))
            msigma[j][k].append(float_str(smass[k]))
            rarr[j][k].append(float_str(rads[k]))

    # Pass as dictionary.
    pars_dict = {
        'gal_names': gal_names, 'ra': ra, 'dec': dec, 'zarr': zarr,
        'zsigma': zsigma, 'aarr': aarr, 'asigma': asigma, 'earr': earr,
        'esigma': esigma, 'darr': darr, 'dsigma': dsigma, 'marr': marr,
        'msigma': msigma, 'rarr': rarr, 'ext_sf': ext_sf, 'ext_mcev': ext_mcev,
        'rad_pc': rad_pc, 'erad_pc': erad_pc, 'int_colors': int_colors,
        'n_memb': n_memb, 'cont_ind': cont_ind, 'kde_prob': kde_prob,
        'r_core_pc': r_core_pc, 'e_r_core': e_r_core, 'phot_disp': phot_disp
    }

    return pars_dict
